crates/glaredb/src/lib.rs

```diff
@@ -0,0 +1,396 @@
+                Err(_) => Some(PathBuf::from(s)),
+            },
+            None => None,
+        }
+    }
+
+    fn cloud_url(&self) -> Option<Url> {
+        self.connection_target
+            .clone()
+            .and_then(|v| Url::parse(&v).ok())
+    }
+}
+
+#[derive(Clone)]
+pub struct Connection {
+    session: Arc<Mutex<TrackedSession>>,
+    _engine: Arc<Engine>,
+}
+
+impl Connection {
+    pub fn execute(&self, query: impl Into<String>) -> Operation {
+        Operation {
+            op: OperationType::Execute,
+            query: query.into(),
+            conn: Arc::new(self.clone()),
+            schema: None,
+        }
+    }
+
+    pub fn sql(&self, query: impl Into<String>) -> Operation {
+        Operation {
+            op: OperationType::Sql,
+            query: query.into(),
+            conn: Arc::new(self.clone()),
+            schema: None,
+        }
+    }
+
+    pub fn prql(&self, query: impl Into<String>) -> Operation {
+        Operation {
+            op: OperationType::Prql,
+            query: query.into(),
+            conn: Arc::new(self.clone()),
+            schema: None,
+        }
+    }
+}
+
+pub struct RecordStream(Pin<Box<dyn Stream<Item = Result<RecordBatch, DataFusionError>> + Send>>);
+
+impl Stream for RecordStream {
+    type Item = Result<RecordBatch, DataFusionError>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        self.0.poll_next_unpin(cx)
+    }
+}
+
+impl From<SendableRecordBatchStream> for RecordStream {
+    fn from(val: SendableRecordBatchStream) -> Self {
+        RecordStream(val.boxed())
+    }
+}
+
+impl From<Result<SendableRecordBatchStream, DataFusionError>> for RecordStream {
+    fn from(val: Result<SendableRecordBatchStream, DataFusionError>) -> Self {
+        match val {
+            Ok(stream) => stream.into(),
+            Err(err) => RecordStream(Operation::handle_error(err).boxed()),
+        }
+    }
+}
+
+impl From<Result<SendableRecordBatchStream, ExecError>> for RecordStream {
+    fn from(val: Result<SendableRecordBatchStream, ExecError>) -> Self {
+        match val {
+            Ok(stream) => stream.into(),
+            Err(err) => RecordStream(Operation::handle_error(err).boxed()),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+enum OperationType {
+    Sql,
+    Prql,
+    Execute,
+}
+
+#[derive(Clone)]
+#[must_use = "operations do nothing unless call() or execute() run"]
+pub struct Operation {
+    op: OperationType,
+    query: String,
+    conn: Arc<Connection>,
+    schema: Option<Arc<Schema>>,
+}
+
+impl Debug for Operation {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "Operation<{:?}>(\"{:?}\")", self.op, self.query)
+    }
+}
+
+impl Operation {
+    pub fn schema(&self) -> Option<Arc<Schema>> {
+        self.schema.clone()
+    }
+
+    pub async fn execute(&mut self) -> Result<SendableRecordBatchStream, ExecError> {
```

Comment:
What's the rationale for `&mut self` instead of `&self`?

Personally I'd expect an operation to be executed (or called) at most once, but I don't feasibility of this accepting a `self`.

Reply:
It was mutable, because of the schema thing, which was for an earlier iteration where the operation itself was a stream.

I've addressed that part of this, leaving the "re-executable" as a discussion (we can also rename `Output` to something else, as I think the naming is just bad, `Execution` makes sense, but that has the word "execute" three times in a single options which seems beyond the pail too.

Reply:
ok, having tried to unwind this the train of thought is:
- we need to be able to treat operations as tables, and implement table provider. the requirements of this functionality require:
  - the operation to know what the output schema is (lazily)
  - be re-runable in case it's reused in different operations.
- the partition wrapper needs to be sync, and so we can't just pass the stream in directly to it.

This explains all of the fraught Arc/Mutex shenanigans.
- the operation needs to be mutable so that we can set schema when the operation evaluates not when it's called.
- the table object needs to be sync

All of this adds up to the current implementation.